pkg/controllers/resources/volumesnapshots/volumesnapshotcontents/syncer.go

```diff
@@ -0,0 +1,166 @@
+		virtualClient:   ctx.VirtualManager.GetClient(),
+		localClient:     ctx.LocalManager.GetClient(),
+		translator:      translate.NewDefaultClusterTranslator(ctx.Options.TargetNamespace, nameTranslator),
+	})
+}
+
+var _ generic.BackwardSyncer = &syncer{}
+
+type syncer struct {
+	generic.Translator
+
+	targetNamespace string
+	virtualClient   client.Client
+	localClient     client.Client
+
+	translator translate.Translator
+}
+
+func NewVolumeSnapshotContentTranslator(physicalNamespace string) translate.PhysicalNameTranslator {
+	return func(vName string, vObj client.Object) string {
+		return translateVolumeSnapshotContentName(physicalNamespace, vName)
+	}
+}
+
+func (s *syncer) New() client.Object {
+	return &volumesnapshotv1.VolumeSnapshotContent{}
+}
+
+func (s *syncer) Backward(ctx context.Context, pObj client.Object, log loghelper.Logger) (ctrl.Result, error) {
+	pVSC := pObj.(*volumesnapshotv1.VolumeSnapshotContent)
+	// check if the VolumeSnapshotContent should get synced
+	sync, vVS, err := s.shouldSync(ctx, pVSC)
+	if err != nil {
+		return ctrl.Result{}, err
+	} else if !sync {
+		// ignore this VolumeSnapshotContent resource, because there is no virtual VolumeSnapshot bound to it"
+		return ctrl.Result{}, nil
+	}
+
+	vVSC := s.translateBackwards(pVSC, vVS)
+	log.Infof("create VolumeSnapshotContent %s, because it does not exist in the virtual cluster", vVSC.Name)
+	return ctrl.Result{}, s.virtualClient.Create(ctx, vVSC)
+}
+
+func (s *syncer) Forward(ctx context.Context, vObj client.Object, log loghelper.Logger) (ctrl.Result, error) {
+	// We are not doing any syncing Forward for the VolumeSnapshotContentes
+	// if this method is called it means that VolumeSnapshotContent was deleted in host,
+	// or a new VolumeSnapshotContent was created in vcluster, and it should be deleted.
+
+	//remove finalizers if present on vObj
+	if len(vObj.GetFinalizers()) > 0 {
+		vObj.SetFinalizers([]string{})
+		err := s.virtualClient.Update(ctx, vObj)
+		if err != nil {
+			return ctrl.Result{}, err
+		}
+	}
+	log.Infof("delete VolumeSnapshotContent %s, because it does not exist in the host cluster", vObj.GetName())
+	zero := int64(0)
+	return ctrl.Result{}, s.virtualClient.Delete(ctx, vObj, &client.DeleteOptions{
+		GracePeriodSeconds: &zero,
+	})
+}
+
+func (s *syncer) Update(ctx context.Context, pObj client.Object, vObj client.Object, log loghelper.Logger) (ctrl.Result, error) {
+	pVSC := pObj.(*volumesnapshotv1.VolumeSnapshotContent)
+	vVSC := vObj.(*volumesnapshotv1.VolumeSnapshotContent)
+
+	//check if objects are getting deleted
+	if pObj.GetDeletionTimestamp() != nil {
+		if vObj.GetDeletionTimestamp() == nil {
+			log.Infof("delete virtual VolumeSnapshotContent %s, because physical VolumeSnapshotContent is being deleted", vObj.GetName())
+			err := s.virtualClient.Delete(ctx, vObj, &client.DeleteOptions{
+				GracePeriodSeconds: pObj.GetDeletionGracePeriodSeconds(),
+			})
+			if err != nil {
+				return ctrl.Result{}, err
+			}
+		}
```

Comment:
We should terminate here, shouldn't we?

Reply:
No, we are checking that multiple Kinds exist so we should let it loop through all of them. :)

Reply:
Pretty much yes :)
I just added extra syncing of finalizers and status to allow for troubleshooting in case the deletion doesn't go smoothly, and I return/terminate afterwards.